002-es-modules.md

```diff
@@ -0,0 +1,175 @@
+  - This, however, is not an option under the RMR.
+4. Conditional imports & exports are impossible.
+  - (The RMR must be constructed at parse time.)
+  - This renders writing Native Modules in ESM impossible.
+
+### 1.2. Avoidance of Async Module Resolution
+
+1. Allows modules to both use Module Mode and be `require()`'d as usual.
+  - Prevents any further interop issues related to timing.
+  - Allows existing modules to transition to NM2 while dependants can continue using NCJS to require them without any changes.
+
+## 2. Related
+
+[ECMA262](tc39.github.io/ecma262/) discusses the syntax and semantics of
+related syntax, and introduces:
+
+### 2.1. Types
+
+* **[ModuleRecord]
+(https://tc39.github.io/ecma262/#sec-abstract-module-records)**
+    - Defines the list of imports via `[[ImportEntry]]`.
+    - Defines the list of exports via `[[ExportEntry]]`.
+
+* **[ModuleNamespace]
+(https://tc39.github.io/ecma262/#sec-module-namespace-objects)**
+    - Represents a read-only static set of bindings to a module's exports.
+
+### 2.2. Operations
+
+* **[ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)**
+    - Creates a [SourceTextModuleRecord]
+    (https://tc39.github.io/ecma262/#sec-source-text-module-records) from
+    source code.
+
+* **[HostResolveImportedModule]
+(https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)**
+    - A hook for when an import is exactly performed. This returns a
+      `ModuleRecord`. Used as a means to grab modules from Node's loader/cache.
+
+* **[ModuleNamespaceCreate]
+(https://tc39.github.io/ecma262/#sec-modulenamespacecreate)**
+    - Provides a means of creating a list of exports manually, used so that
+      NCJS `module.exports` can create `ModuleRecord`s that are
+      prepopulated.
+
+## 3. Semantics
+
+### 3.1. Determining if source is an ES Module
+
+A new file type will be used for Node v2.0 Modules, `.mjs`. Files with this extension
+will be treated using the loading semantics in this document, but retain a way
+to load existing extensions of `.json`, `.node`, and `.js`. The file type for
+JavaScript with IANA as an official file type needs to have this extensions
+added. This comes from browsers being unable to implement a new MIME for the
+differing parse goals. In order to make this change, contact the IESG.
+
+The `.mjs` file extension will be searched for prior to any `.js` file in any
+algorithm that uses searching (`require`); e.g. once the Node
+resolution algorithm reaches file expansion, `path + '.mjs'` would be searched
+prior to `path + '.js'` when performing `require(path)`.
+
+# 3.1.1. Inter package loading using file extension breakage.
+
+There is knowledge of breakage for code that upgrades inner package
+dependencies such as `require('foo/bar.js')`. As `bar.js` may move to
+`bar.mjs`. Since `bar.js` is not the listed entry point this was considered
+acceptable.
+
+### 3.2. Importing Modules in NM2
+
+Node v2.0 Modules do not implement `import` and `export`.
```

Comment:
Another point - at some point, sooner than later, new JS API features may *only* be made available via `import` - ie, importing "builtin" modules. If `import` is excluded, then `node` would have to do something like overload `require` to support builtin modules, which has its own problems.